### DISCUSSION TITLE
Route group and layout at top level prevents root index page from being rendered

### DISCUSSION THREAD

**Discussion by octopicorn** [TanStack/router#1457] [General]
With the following file structure (file based routing):

```text
src/routes/
  (landing-pages)/  <- this is a route group which will not appear in the url path
      _layout/
          index.tsx // <- this is the home page with layout applied
          about.tsx
          contact.tsx
      _layout.tsx // <- this is the layout file
```
 
 
 
 Expected result:
 Home page is rendered with the layout applied.
    
 Actual result:
 No home page is rendered at all.
    
 If I remove the (landing-pages) route group, rendering works again as expected.  There seems to be something about the route group at top level that prevents the route '/' from rendering _layout/index.tsx wrapped by _layout.tsx.
    
I've reproduced this in a [CodePen](https://codesandbox.io/p/devbox/github/tanstack/router/tree/main/examples/react/basic-file-based?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluukm00s00073b6l5ld3v33r%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluukm00s00023b6ls0qyrkb1%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluukm00s00043b6l3nfbc77b%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluukm00s00063b6l0wkm2ovs%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluukm00s00023b6ls0qyrkb1%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluukm00r00013b6l1tyi15hs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cluukm00s00023b6ls0qyrkb1%2522%252C%2522activeTabId%2522%253A%2522cluukm00r00013b6l1tyi15hs%2522%257D%252C%2522cluukm00s00063b6l0wkm2ovs%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluukm00s00053b6ll7bwo0sq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3001%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cluukm00s00063b6l0wkm2ovs%2522%252C%2522activeTabId%2522%253A%2522cluukm00s00053b6ll7bwo0sq%2522%257D%252C%2522cluukm00s00043b6l3nfbc77b%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluukm00s00033b6ldczya4m8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522cluukm00s00043b6l3nfbc77b%2522%252C%2522activeTabId%2522%253A%2522cluukm00s00033b6ldczya4m8%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) based off of the Basic example in the docs.
    
Is layout supposed to work in combination with route groups? Or is it an either/or set of features?

This is a real use case. I want Home and several other pages to share a "top level" layout, let's say a left nav and header. Then I want a set of subpages to share a completely different layout (not inheriting from top level layout at all), like a project workspace with tabs.

I thought route groups might offer me a way to separate these two groups of layout groups, but it seems like there is something about the root index page that doesn't fit this pattern.

**Comment by schiller-manuel** [comment]
looks like a bug in the generator. can you please create an issue with a minimal complete example (fork one of the existing examples on Stackblitz)?

**Comment by CanRau** [comment]
Also trying to have root index nested in a layout to completely separate it from auth & other layouts

Tried couple of ways

```sh
tree app/routes
app/routes
├── (auth)
│   └── login.tsx
├── (main)
│   ├── _layout
│   │   └── index.tsx
│   └── _layout.tsx
└── __root.tsx
```

```sh
tree app/routes
app/routes
├── (auth)
│   └── login.tsx
├── (main)
│   └── index.tsx
├── _(main).tsx
└── __root.tsx
```

```sh
tree app/routes
app/routes
├── (auth)
│   └── login.tsx
├── _main
│   └── index.tsx
├── _main.tsx
└── __root.tsx
```

Actually while writing this I got the 1. approach kinda working in Stackblitz tho navigating causes different errors https://stackblitz.com/edit/github-3ber8h locally I can't get it working at all

Also played around with the `routeToken` approach using the default and changing it to `layout` as suggested here https://github.com/TanStack/router/discussions/1102#discussioncomment-10946603 tho still failing

**Comment by CanRau** [comment]
Tried this one as it looks promising https://github.com/SeanCassiere/nv-rental-clone/blob/master/src/routes/_auth/(dashboard)/_index.tsx

with the following tsr settings

```
indexToken: "_index",
routeToken: "_route",
```

tho I get the following error when navigating to any route within `_main` where `_index.tsx` lives in my case. `routeTree.gen` looks good to me

```sh
Error in renderToPipeableStream: TypeError: String.prototype.search called on null or undefined
    at search (<anonymous>)
    at functionalUpdate (file:///[..]/node_modules/.pnpm/@tanstack+react-router@1.81.10_@tanstack+router-generator@1.81.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@tanstack/react-router/dist/esm/utils.js:10:12)
    at final (file:///[..]/node_modules/.pnpm/@tanstack+react-router@1.81.10_@tanstack+router-generator@1.81.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@tanstack/react-router/dist/esm/router.js:429:20)
    at applyNext (file:///[..]/node_modules/.pnpm/@tanstack+react-router@1.81.10_@tanstack+router-generator@1.81.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@tanstack/react-router/dist/esm/router.js:440:20)
    at applyMiddlewares (file:///[..]/node_modules/.pnpm/@tanstack+react-router@1.81.10_@tanstack+router-generator@1.81.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@tanstack/react-router/dist/esm/router.js:442:18)
    at build (file:///[..]/node_modules/.pnpm/@tanstack+react-router@1.81.10_@tanstack+router-generator@1.81.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@tanstack/react-router/dist/esm/router.js:444:18)
    at buildWithMatches (file:///[..]/node_modules/.pnpm/@tanstack+react-router@1.81.10_@tanstack+router-generator@1.81.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@tanstack/react-router/dist/esm/router.js:463:22)
    at Router.buildLocation (file:///[..]/node_modules/.pnpm/@tanstack+react-router@1.81.10_@tanstack+router-generator@1.81.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@tanstack/react-router/dist/esm/router.js:504:14)
```
this is just an exerpt the full trace is quite long

**Reply by schiller-manuel** [reply]
you cannot use a _ in the token as they will be interpreted as layouts

**Reply by CanRau** [reply]
removed token config and renamed to `index.tsx` still same error

**Reply by schiller-manuel** [reply]
can you please create a minimal router-only example and create a new issue for this?

**Reply by CanRau** [reply]
actually can't reproduce anymore ^^
not sure it's because of the latest updates or something else in repo 🤷🏼 

Update: yea something with my repo, just started a new one, even patched my PR #2734 in and having `index.tsx` nested within pathless routes & groups works just fine 😅 

tested examples 1 & 3 (https://github.com/TanStack/router/discussions/1457#discussioncomment-11262761) and even a variation of 3 where `index.tsx` within `_main` is nested in a pathless group `(inner)`

**Reply by schiller-manuel** [reply]
great, so we can close this? :D

**Reply by CanRau** [reply]
At least regarding my stuff yes, not sure about OP

**Reply by QingjiaTsang** [reply]
I came across the same problem as the OP faced. And also, I see your first approach above:
```
tree app/routes
app/routes
├── (auth)
│   └── login.tsx
├── (main)
│   ├── _layout
│   │   └── index.tsx
│   └── _layout.tsx
└── __root.tsx
```

Have you found a way to make it work?

I want to use the route group to nest the landing page route files, and in my case, I tried 3 ways:
1. This is the ideal way but failed. It turns out to be not rendered anything when it comes to "/" route just like the OP said:
<img width="333" alt="image" src="https://github.com/user-attachments/assets/a60f7bfa-4454-4f48-a534-51d0c660e814" />

or 2. I also tried another way to apply the layout and it's still not working with route group:
<img width="337" alt="image" src="https://github.com/user-attachments/assets/51d205de-fc67-485f-8acb-4de639ea2e0c" />


3. Works if I don't use route group:
<img width="361" alt="image" src="https://github.com/user-attachments/assets/236b3f31-26ca-4fe0-aa29-bd7f399dc968" />

I don't know whether this issue has solved for now. I'll be appreciated if you share any solutions.

**Reply by schiller-manuel** [reply]
@QingjiaTsang can you provide a minimal complete example?

**Reply by QingjiaTsang** [reply]
@schiller-manuel https://github.com/QingjiaTsang/tanstack-router-route-groups-issue

<img width="492" alt="image" src="https://github.com/user-attachments/assets/b2d3f4fb-36dc-4970-b604-e38e57148b4f" />
<img width="1956" alt="image" src="https://github.com/user-attachments/assets/43ba5877-6f6a-4053-9933-4066833c6335" />

Eventually, I found out that it has something to do with the other pathless route inside another route group like the ~_authenticated-layout.tsx file in the pic. In this case, if you delete the ~_authenticated-layout.tsx file, things will go well.

**Reply by QingjiaTsang** [reply]
@schiller-manuel Did I make it clear?

**Reply by 1hachem** [reply]
I am facing the same issue now, did you find any fix for this @QingjiaTsang

**Reply by 1hachem** [reply]
for now my solution is to move the `/` and its layout to the root of my `routes` directory (moved it out of the `(landing)/` group)
